### PR TITLE
Implementation for Atlas Scientific Peristaltic Pump

### PR DESCRIPTION
**Pull request by carlos-sarmiento**
# What does this implement/fix?

Initial implementation of most of the functionality from Atlas Scientific's Peristaltic Pump (https://atlas-scientific.com/peristaltic/ezo-pmp/)

Datasheet: https://files.atlas-scientific.com/EZO_PMP_Datasheet.pdf

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2115

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
i2c:

ezo_pmp:
  id: hcl_pump
  update_interval: 1s

binary_sensor:
  - platform: ezo_pmp
    pump_state:
      name: "Pump State"
    is_paused:
      name: "Is Paused"

sensor:
  - platform: ezo_pmp
    current_volume_dosed:
      name: "Current Volume Dosed"
    total_volume_dosed:
      name: "Total Volume Dosed"
    absolute_total_volume_dosed:
      name: "Absolute Total Volume Dosed"
    pump_voltage:
      name: "Pump Voltage"
    last_volume_requested:
      name: "Last Volume Requested"
    max_flow_rate:
      name: "Max Flow Rate"

text_sensor:
  - platform: ezo_pmp
    dosing_mode:
      name: "Dosing Mode"
    calibration_status:
      name: "Calibration Status"

button:

  - platform: template
    name: "Dose Volume"
    on_press:
      - ezo_pmp.dose_volume:
          id: hcl_pump
          volume: 10

  - platform: template
    name: "Dose Volume/Time"
    on_press:
      - ezo_pmp.dose_volume_over_time:
          id: hcl_pump
          volume: 10
          duration: 2

  - platform: template
    name: "Dose Constant Flow Rate"
    on_press:
      - ezo_pmp.dose_with_constant_flow_rate:
          id: hcl_pump
          volume_per_minute: 10
          duration: 2

  - platform: template
    name: "Set Calibration Volume"
    on_press:
      - ezo_pmp.set_calibration_volume:
          id: hcl_pump
          volume:10

  - platform: template
    name: "Find"
    on_press:
      - ezo_pmp.find: hcl_pump
  - platform: template
    name: "Dose Continuously"
    on_press:
      - ezo_pmp.dose_continuously: hcl_pump
  - platform: template
    name: "Clear Total Volume Closed"
    on_press:
      - ezo_pmp.clear_total_volume_dosed: hcl_pump
  - platform: template
    name: "Clear Calibration"
    on_press:
      - ezo_pmp.clear_calibration: hcl_pump
  - platform: template
    name: "Pause/Unpause Dosing"
    on_press:
      - ezo_pmp.pause_dosing: hcl_pump
  - platform: template
    name: "Stop Dosing"
    on_press:
      - ezo_pmp.stop_dosing: hcl_pump

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ezo_pmp/__init__.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        ),  # Any way to represent as proper volume (vs. raw int)
+    }
+)
+
+
+@automation.register_action(
+    "ezo_pmp.dose_volume", EzoPMPDoseVolumeAction, EZO_PMP_DOSE_VOLUME_ACTION_SCHEMA
+)
+async def ezo_pmp_dose_volume_to_code(config, action_id, template_arg, args):
+    paren = await cg.get_variable(config[CONF_ID])
```
Comment (PoppyPop) [inline esphome/components/ezo_pmp/__init__.py:150]: Maybe a more explicit variable name could be more readable. Made the comment one time, but this is for the entire file

**Review by PoppyPop — COMMENTED** [review]
LGTM, some minor comment. 
I have tested it with my pump work great.
Thank you

File: esphome/components/ezo_pmp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+EZO_PMP_DOSE_VOLUME_OVER_TIME_ACTION_SCHEMA = cv.All(
+    {
+        cv.Required(CONF_ID): cv.use_id(EzoPMP),
+        cv.Required(CONF_VOLUME): cv.templatable(
+            cv.float_range()
+        ),  # Any way to represent as proper volume (vs. raw int)
+        cv.Required(CONF_DURATION): cv.templatable(
+            cv.int_range(1)
+        ),  # Any way to represent it as minutes (vs. raw int)
+    }
```
Comment (jesserockz) [inline esphome/components/ezo_pmp/__init__.py:168]: See `cv.float_with_unit` and `cv.positive_time_period_minutes`

File: esphome/components/ezo_pmp/__init__.py (hunk 1 above)
Comment (carlos-sarmiento) [reply esphome/components/ezo_pmp/__init__.py:150]: I don't understand what variable name we are talking about here

File: esphome/components/ezo_pmp/__init__.py (hunk 2 above)
Comment (carlos-sarmiento) [reply esphome/components/ezo_pmp/__init__.py:168]: Will update for this

File: esphome/components/ezo_pmp/__init__.py (hunk 1 above)
Comment (PoppyPop) [reply esphome/components/ezo_pmp/__init__.py:150]: `paren`: i think adding the missing t can accelerate reading, it take me some reflexion if you will say parent or misspelled another word.

Like i said nothing blocking.

**Comment by Spaceinvaderz** [comment]
@carlos-sarmiento Hi! Any hints how to use this component if you have more than 1 EZO Pump? 
Any config examples would be great.

**Comment by carlos-sarmiento** [comment]
You can define several records on the `ezo_pmp` key with a different ID and the address for each pump

**Comment by Spaceinvaderz** [comment]
> 

@carlos-sarmiento  Looks like it works for me, thank you! 

And what about sensors? I.e.:  

```yaml
binary_sensor:
  - platform: ezo_pmp
    pump_state:
      name: "Pump State"
    is_paused:
      name: "Is Paused"

sensor:
  - platform: ezo_pmp
    current_volume_dosed:
      name: "Current Volume Dosed"
    total_volume_dosed:
      name: "Total Volume Dosed"
    absolute_total_volume_dosed:
      name: "Absolute Total Volume Dosed"
    pump_voltage:
      name: "Pump Voltage"
    last_volume_requested:
      name: "Last Volume Requested"
    max_flow_rate:
      name: "Max Flow Rate"

text_sensor:
  - platform: ezo_pmp
    dosing_mode:
      name: "Dosing Mode"
    calibration_status:
```
In case of sensors I can not make it work in multi pump config,

**Comment by carlos-sarmiento** [comment]
You would need to include the id of the pump on the definition of each sensor

**Review by jesserockz — APPROVED** [review]

**Pull request merged**